feat: add cache verification tool for dashbaord

RedemptionC opened the pull request:
Please answer these questions before submitting a pull request, **or your PR will get closed**.

**Why submit this pull request?**

- [ ] Bugfix
- [x] New feature provided
- [ ] Improve performance
- [ ] Backport patches

**What changes will this PR take into?**
* add a new API to verify that items in cache are the same as those in ETCD
* add a new cobra command to do the same stuff,by calling this API
* I also changed authentication.go to temporarily bypass the token verification(to use this API without login)

this is first vesion of this tool,so it doesn't include any test or doc,I need to request some advice on implementation to find the correct way to implement it
and I will add test and doc(if necessary) soon
**Related issues**
fix #2063 

**Checklist:**

- [x] Did you explain what problem does this PR solve? Or what new features have been added?
- [x] Have you added corresponding test cases?
- [x] Have you modified the corresponding document?
- [ ] Is this PR backward compatible? If it is not backward compatible, please discuss on the mailing list first

RedemptionC committed:
feat: add cache verification tool for dashboard (#2063)

RedemptionC committed:
fix: handle error when closing response body

RedemptionC committed:
fix: add license header

RedemptionC commented:
2 E2E test failed: front-end and back-end
I have seen the back-end one in previous PR,it seems to be unrelated to our modifications,see https://github.com/apache/apisix-dashboard/pull/1985#event-5072534144
does anyone know how to fix or bypass them?

Baoyuantop commented:
> 2 E2E test failed: front-end and back-end
> I have seen the back-end one in previous PR,it seems to be unrelated to our modifications,see [#1985 (comment)](https://github.com/apache/apisix-dashboard/pull/1985#event-5072534144)
> does anyone know how to fix or bypass them?

Maybe you can merge the latest master branch code?

RedemptionC commented:
> > 2 E2E test failed: front-end and back-end
> > I have seen the back-end one in previous PR,it seems to be unrelated to our modifications,see [#1985 (comment)](https://github.com/apache/apisix-dashboard/pull/1985#event-5072534144)
> > does anyone know how to fix or bypass them?
> 
> Maybe you can merge the latest master branch code?

Hello,I am using the lastest master code as base

Baoyuantop commented:
> > > 2 E2E test failed: front-end and back-end
> > > I have seen the back-end one in previous PR,it seems to be unrelated to our modifications,see [#1985 (comment)](https://github.com/apache/apisix-dashboard/pull/1985#event-5072534144)
> > > does anyone know how to fix or bypass them?
> > 
> > 
> > Maybe you can merge the latest master branch code?
> 
> Hello,I am using the lastest master code as base

let me check

Baoyuantop commented:
Similar to #2059

RedemptionC commented:
> Similar to #2059

yes,the front-end part is the same
But I did use your patch,why did it failed?

RedemptionC committed:
fix: increment timeout value to pass front-end E2E Test

RedemptionC commented:
@nic-chen

nic-chen commented on api/internal/filter/authentication.go:
```diff
@@ -45,7 +45,7 @@ func (mw *AuthenticationMiddleware) Handle(ctx droplet.Context) error {
 	req := httpReq.(*http.Request)
 
-	if req.URL.Path == "/apisix/admin/tool/version" || req.URL.Path == "/apisix/admin/user/login" {
+	if req.URL.Path == "/apisix/admin/tool/version" || req.URL.Path == "/apisix/admin/user/login" || req.URL.Path == "/apisix/admin/cache_verify" {
```
need auth for the new API

RedemptionC replied:
what's the proper way to use this API in cache_verify command?
is the current implementation desirable?
so we just need to first get token,then use this API this way?

nic-chen commented on api/cmd/cache_verify.go:
```diff
@@ -0,0 +1,49 @@
+		Use:   "cache-verify",
+		Short: "verify that data in cache are consistent with that in ETCD",
+		Run: func(cmd *cobra.Command, args []string) {
+			rsp, err := http.Get("http://localhost:9000/apisix/admin/cache_verify")
```
the port is not fixed

RedemptionC replied:
how do we know which port is manager API running on?
we can not use conf.ServerPort,as it will be 80(the default value) when we run cache_verify command
should we read the configuration file?but it will result in  quite a few duplicate code
can we ask user to pass them as command line arguments?

nic-chen commented on api/internal/handler/cache_verify/cache_verify.go:
```diff
@@ -0,0 +1,155 @@
+}
+
+type compareResult struct {
+	K          string `json:"key"`
```
K —> Key

nic-chen commented on api/internal/handler/cache_verify/cache_verify.go:
```diff
@@ -0,0 +1,155 @@
+	}
+	cacheValue := string(s)
+	key := fmt.Sprintf("/apisix/%s/%s", infix, k)
+	val, err := storage.GenEtcdStorage().Get(context.TODO(), key)
```
We could get it in batches by prefix, not every key to get once.

RedemptionC replied:
> get it in batches by prefix

what does this mean? 
does it mean that we should get etcd client in the CacheVerify function,and do it once every time when we run this command ?

nic-chen replied:
See the difference between `etcdctl get  --prefix /apisix/routes` and `etcdctl get /apisix/routes/1`

RedemptionC replied:
> See the difference between `etcdctl get --prefix /apisix/routes` and `etcdctl get /apisix/routes/1`

got it 
I will change it in next commit

nic-chen commented on web/cypress/integration/route/search-route.spec.js:
```diff
@@ -17,7 +17,7 @@
 
 context('Create and Search Route', () => {
-  const timeout = 500;
+  const timeout = 5000;
```
why change this？

RedemptionC replied:
>Similar to #2059

see this,we can't pass front-end test in CI without changing this

nic-chen reviewed (commented):
need test cases

RedemptionC committed:
chore:
1. change variable name to make it clear
2. get ETCD client every time we call this API,rather every time an item is retrieved from ETCD

RedemptionC committed:
Revert "fix: increment timeout value to pass front-end E2E Test"

This reverts commit ece7148ddb9b0ffc137f1584d21382a9c0b896ca.

RedemptionC committed:
fix: rewrite cache_verify API using etcd.List(ETCD get with prefix)

RedemptionC committed:
fix: simplify code

RedemptionC committed:
fix:
1. require token to use cache_verify API
2. add simple test

RedemptionC committed:
fix: add license

RedemptionC committed:
fix: typo

RedemptionC committed:
fix: delete meaningless test

RedemptionC committed:
fix: add test using ginkgo

RedemptionC committed:
fix: read serverPort from conf file in cache_verify command

RedemptionC committed:
fix: add license

RedemptionC committed:
fix: rename struct for cache_verify

RedemptionC committed:
chore: add doc for cache verify API

RedemptionC committed:
chore: format markdown

RedemptionC committed:
chore: format markdown

RedemptionC committed:
chore: format output of cache verify command

RedemptionC committed:
fix: go lint

RedemptionC committed:
fix: add more test for cache verify

RedemptionC commented:
in the test file,do we need to check whether the result is correct by comparing cache and ETCD again?
@nic-chen

nic-chen commented:
@RedemptionC 

I think the response body should also contain detailed information about consistent data, including how much data is consistent for each resource. And check them in e2e test case.

RedemptionC commented:
> I think the response body should also contain detailed information about consistent data, including how much data is consistent for each resource. 

@nic-chen 
currently,when nothing is inconsistent,the response of this API and the ouput of the cache-verify is as follows:
```sh
❯ curl -s localhost:9000/apisix/admin/cache_verify -H "Authorization: $(curl -s  localhost:9000/apisix/admin/user/login -X POST -d '{"username":"admin","password":"admin"}'  | jq -r '.data | .token')" | jq
{
  "code": 0,
  "message": "",
  "data": {
    "inconsistent_consumers": null,
    "inconsistent_routes": null,
    "inconsistent_services": null,
    "inconsistent_ssls": null,
    "inconsistent_upstreams": null,
    "inconsistent_scripts": null,
    "inconsistent_global_plugins": null,
    "inconsistent_plugin_configs": null,
    "inconsistent_server_infos": null
  },
  "request_id": "899e1ffb-0b74-4ecb-93ae-4c19f320b6b9"
}

```
```sh
❯ go run main.go cache-verify 
cache verification result as follows:

ssls            :       all consistent
routes          :       all consistent
scripts         :       all consistent
services        :       all consistent
upstreams       :       all consistent
consumers       :       all consistent
server infos    :       all consistent
global plugins  :       all consistent
plugin configs  :       all consistent
```

do you think that we should maintain some statistical data for each item,like:
"there are 10 consumers in cache,4 of them are consistent,6 of them are inconsistent."
and for each inconsistent item,we provide their content in both cache and etcd?

RedemptionC committed:
fix: modify handler/cache_verify to get ready for dependency injection(so we can mock)

RedemptionC committed:
fix: draft of table driven test(with mock)

RedemptionC committed:
fix: modify API response and add unit test using mock

RedemptionC commented:
> @RedemptionC
> 
> I think the response body should also contain detailed information about consistent data, including how much data is consistent for each resource. And check them in e2e test case.
> 
> And you could also use mock data for unit test.

@nic-chen 
Hello,I have changed something as we discussed above
maybe you could have a look
If it's OK,I will polish them and add detailed unit tests and e2e test soon

RedemptionC committed:
fix:
1. fix go-lint
2. add license
3. delete e2e test for cache verify for now

RedemptionC committed:
fix:
update  doc for cache verify API

RedemptionC commented:
@nic-chen 
but the command line output is pretty ugly....
```
cache verification result as follows:

There are 1 items in total,0 of them are consistent,1 of them are inconsistent
ssls           : 0 in total,0 consistent,0 inconsistent
routes         : 0 in total,0 consistent,0 inconsistent
scripts        : 0 in total,0 consistent,0 inconsistent
services       : 0 in total,0 consistent,0 inconsistent
upstreams      : 0 in total,0 consistent,0 inconsistent
consumers      : 1 in total,0 consistent,1 inconsistent
inconsistent consumers:
[key](andy)
[etcd]({"username":"andy","plugins":{"key-auth":{"key":"key-of-john"}},"create_time":1627739045,"update_time":1627744978})
[cache]({"create_time":1627739046,"plugins":{"key-auth":{"key":"key-of-john"}},"update_time":1627744978,"username":"andy"})
server infos   : 0 in total,0 consistent,0 inconsistent
global plugins : 0 in total,0 consistent,0 inconsistent
plugin configs : 0 in total,0 consistent,0 inconsistent
```
do you have any advice?

RedemptionC committed:
fix:
fix markdown=lint:consecutive blank line

RedemptionC committed:
fix:
add simple e2e test for cache_verify

RedemptionC commented:
@nic-chen 
how to set up the init state of testing environment?
when I'm testing this API in e2e testing,I created 3 items,I wanted to assert the total number of items is 3,but actually the error log in github actions show that there are 5
```
Cache verify [It] cache verify  
/home/runner/work/apisix-dashboard/apisix-dashboard/api/test/e2enew/cache_verify/cache_verify_test.go:31

  Expected
      <int64>: 5
  to equal
      <int64>: 3
```
I'm confused.What should I test in e2e tesing? It seems that  unit testing using mock works best in this case,while it's hard to test the API's functionality using e2e testing
what do you think?

RedemptionC committed:
fix:modify e2e test for cache_verify:

assert the diff between 2 API response

RedemptionC commented:
@nic-chen 
maybe we can assert the diff? like my latest commit did?
is the test now is  thorough enough?
and,what kind of format of the command line output is desirable?

RedemptionC commented:
What's your opinion? @starsz @tokers @gxthrj @bzp2010 @imjoey @Jaycean @ShiningRush

RedemptionC commented:
> maybe we can assert the diff? like my latest commit did?

but it seems that this is tesing the create API,not cache verify
what part of cache_verify should we test in the e2e test?

nic-chen commented:
> @nic-chen
> but the command line output is pretty ugly....
> 
> ```
> cache verification result as follows:
> 
> There are 1 items in total,0 of them are consistent,1 of them are inconsistent
> ssls           : 0 in total,0 consistent,0 inconsistent
> routes         : 0 in total,0 consistent,0 inconsistent
> scripts        : 0 in total,0 consistent,0 inconsistent
> services       : 0 in total,0 consistent,0 inconsistent
> upstreams      : 0 in total,0 consistent,0 inconsistent
> consumers      : 1 in total,0 consistent,1 inconsistent
> inconsistent consumers:
> [key](andy)
> [etcd]({"username":"andy","plugins":{"key-auth":{"key":"key-of-john"}},"create_time":1627739045,"update_time":1627744978})
> [cache]({"create_time":1627739046,"plugins":{"key-auth":{"key":"key-of-john"}},"update_time":1627744978,"username":"andy"})
> server infos   : 0 in total,0 consistent,0 inconsistent
> global plugins : 0 in total,0 consistent,0 inconsistent
> plugin configs : 0 in total,0 consistent,0 inconsistent
> ```
> 
> do you have any advice?

LGTM

nic-chen commented:
> > maybe we can assert the diff? like my latest commit did?
> 
> but it seems that this is tesing the create API,not cache verify
> what part of cache_verify should we test in the e2e test?

Why it's not cache verify?

RedemptionC commented:
> > > maybe we can assert the diff? like my latest commit did?
> > 
> > 
> > but it seems that this is tesing the create API,not cache verify
> > what part of cache_verify should we test in the e2e test?
> 
> Why it's not cache verify?

@nic-chen 

in the e2e test, I only check that `total` field in the cache verify response did increase by 3(which is the number of configs I created),I didn't check anything about consistency,becacuse I don't know how to make assertions on them: this is not mock,they can be consistent or inconsistent,I don't know what should I expect from them

If you think the current implementation if OK,I can submit a clean version of them,and maybe then we can request more reviewers to see whether they deceide to approve this PR?

RedemptionC committed:
fix: run goimports on them

RedemptionC committed:
fix: using loop to print result

RedemptionC committed:
fix: read password of admin from conf

RedemptionC committed:
fix: using map and loop to avoid writing multiple if and function call

RedemptionC committed:
fix: use host from conf

RedemptionC committed:
fix: when we come across errors,we use errorf and return

RedemptionC committed:
fix: when we come across errors,we use errorf and return

RedemptionC committed:
fix:
1. when we come across errors,we use errorf and return
2. run goimport
3. delete unnecessary lines

RedemptionC committed:
fix:
1. add more asserts in e2e/cache_verify_test

RedemptionC committed:
fix: typo

RedemptionC committed:
fix: remove hard-coded host,turn to use host from conf

nic-chen commented on api/cmd/cache_verify.go:
```diff
@@ -0,0 +1,177 @@
+
+			get, err := http.NewRequest("GET", url, nil)
+			if err != nil {
+				log.Errorf("new http request failed: %s", err)
```
```suggestion
				fmt.Fprintf(os.Stderr, "new http request failed: %s\n", err)
```

nic-chen commented on api/cmd/cache_verify.go:
```diff
@@ -0,0 +1,177 @@
+
+			rsp, err := client.Do(get)
+			if err != nil {
+				log.Errorf("get result from migrate/export failed: %s", err)
```
ditto

nic-chen commented on api/cmd/cache_verify.go:
```diff
@@ -0,0 +1,177 @@
+
+			port = conf.ServerPort
+			host = conf.ServerHost
+			username = "admin"
```
why not use username in `config.yaml`? it can be not `admin`.

RedemptionC replied:
>@nic-chen
if we use account from the conf,should we use admin or user?

there are 2 user in the config file

nic-chen replied:
use the first one. username can be any character.

nic-chen commented on api/cmd/cache_verify.go:
```diff
@@ -0,0 +1,177 @@
+				rs.Data.Items.ServerInfos,
+				rs.Data.Items.GlobalPlugins,
+				rs.Data.Items.PluginConfigs,
+			}
```
why not use a map? and we could put them into the function `printResult`

RedemptionC replied:
because map doesn't guarantee iteration order,I want to make sure the output are sorted by the config name,like:
```
❯ go run main.go cache-verify 
cache verification result as follows:

There are 12 items in total,12 of them are consistent,0 of them are inconsistent
ssls           : 0 in total,0 consistent,0 inconsistent
routes         : 1 in total,1 consistent,0 inconsistent
scripts        : 0 in total,0 consistent,0 inconsistent
services       : 1 in total,1 consistent,0 inconsistent
upstreams      : 2 in total,2 consistent,0 inconsistent
consumers      : 7 in total,7 consistent,0 inconsistent
server infos   : 1 in total,1 consistent,0 inconsistent
global plugins : 0 in total,0 consistent,0 inconsistent
plugin configs : 0 in total,0 consistent,0 inconsistent
```

nic-chen commented on api/internal/handler/cache_verify/cache_verify_test.go:
```diff
@@ -0,0 +1,104 @@
+		})
+	}
+
+}
```
What is the test coverage?

RedemptionC replied:
![image](https://user-images.githubusercontent.com/17563754/131543517-be69be31-8ca0-4ce7-9bb0-2019c5939429.png)

RedemptionC committed:
fix:
1. group global variable together
2. use fprintf instead of errorf in cmd/cache_verify

RedemptionC commented:
@nic-chen @starsz 
Hello,could you review them so that we can see what we we do to get this PR merged?

liuxiran commented:
hi @RedemptionC , please fix the ci problem when you have time, thanks

RedemptionC commented:
hi @liuxiran 
it's fixed now

liuxiran commented:
ping @nic-chen @starsz please have a look again when you have time, thanks a lot

RedemptionC commented:
ping @nic-chen @starsz please have a look again when you have time, thanks a lot

nic-chen commented on api/cmd/cache_verify.go:
```diff
@@ -0,0 +1,178 @@
+			host = conf.ServerHost
+			username = "admin"
+			password = conf.UserList[username].Password
+
```
Unnecessary blank lines.

nic-chen commented on api/cmd/cache_verify.go:
```diff
@@ -0,0 +1,178 @@
+			password = conf.UserList[username].Password
+
+			token := getToken()
+
```
ditto

nic-chen commented on api/cmd/cache_verify.go:
```diff
@@ -0,0 +1,178 @@
+			url := fmt.Sprintf("http://%s:%d/apisix/admin/cache_verify", host, port)
+			client := &http.Client{}
+
+			get, err := http.NewRequest("GET", url, nil)
```
`get` -> `req` or other meaningful variable name.

nic-chen commented on api/cmd/cache_verify.go:
```diff
@@ -0,0 +1,178 @@
+	data, err := json.Marshal(account)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "json marshal failed: %s\n", err)
+		return ""
```
should be better return error, not just ""

nic-chen commented on api/cmd/cache_verify.go:
```diff
@@ -0,0 +1,178 @@
+			log.InitLogger()
+
+			port = conf.ServerPort
+			host = conf.ServerHost
```
why use `conf.ServerHost`? it can be `0.0.0.0` or something else. I think we could use `127.0.0.1` here.

starsz replied:
Couldn't use 127.0.0.1 here. If the manager-api binds an intranet IP.

nic-chen commented on api/internal/handler/cache_verify/cache_verify.go:
```diff
@@ -0,0 +1,268 @@
+}
+
+func (h *Handler) CacheVerify(_ droplet.Context) (interface{}, error) {
+	checkConsistent := func(hubKey store.HubKey, s *store.Interface, rs *OutputResult, etcd *storage.Interface) {
+
+		keyPairs, err := (*etcd).List(context.TODO(), fmt.Sprintf("/apisix/%s/", infixMap[hubKey]))
```
Please see here

nic-chen replied:
prefix `/apisix/` can be changed in conf.yaml, please use `conf.ETCDConfig.Prefix`

nic-chen commented on api/internal/handler/cache_verify/cache_verify.go:
```diff
@@ -0,0 +1,219 @@
+		for i := range keyPairs {
+			key := path.Base(keyPairs[i].Key)
+
+			cacheObj, err := (*s).Get(context.TODO(), key)
```
what is `s`? please use a meaningful variable name

RedemptionC committed:
fix: apply suggestions from code review

RedemptionC commented:
ping @nic-chen @starsz @juzhiyuan @bzp2010 @liuxiran @LiteSun 
please have a look again when you have time, thanks a lot

juzhiyuan commented:
cc @guoqqqi to have a look

guoqqqi commented:
Hi, @RedemptionC  You can modify this in the front-end E2E test case by adding an assertion to ensure that the drawer is closed before clicking Next.
![image](https://user-images.githubusercontent.com/72343596/133892842-9639eb00-08ca-41d9-91eb-799211942101.png)
If you can, update the latest code, or re-run the CI, so I can pinpoint the problem more accurately.

RedemptionC commented:
@guoqqqi 
Hello,thanks for your reply .  I have fetched and rebased on the latest code,hopefully it can pass the CI now
BTW,as @nic-chen said,one PR does one thing,so maybe I should not commit something about front-end in this PR

juzhiyuan commented:
cc @nic-chen for help

nic-chen commented on api/cmd/cache_verify.go:
```diff
@@ -0,0 +1,181 @@
+			log.InitLogger()
+
+			port = conf.ServerPort
+			host = "127.0.0.1"
```
In two cases, if `conf.ServerHost` is `0.0.0.0` or an empty string, then use `127.0.0.1`, otherwise use `conf.ServerHost`

RedemptionC replied:
@nic-chen 
thanks a lot
I've submitted this change,you could check again

RedemptionC committed:
fix: logic about setting host

RedemptionC committed:
fix: logic about setting host

nic-chen commented:
@RedemptionC please fix the CI.

RedemptionC committed:
fix: change test case to avoid conflict

RedemptionC commented:
> @RedemptionC please fix the CI.

The failure doesn't seem to reproduce all the time
what should I do? 
(and the CI seems to take so long to finish

RedemptionC commented:
@nic-chen 
there are only one failure now,and it seems to be unrelated to my commit.
```
#12 227.7 info There appears to be trouble with your network connection. Retrying...
#12 260.8 info There appears to be trouble with your network connection. Retrying...
#12 294.0 error An unexpected error occurred: "https://registry.nlark.com/monaco-editor/download/monaco-editor-0.24.0.tgz?cache=0&sync_timestamp=1620845835330&other_urls=https%3A%2F%2Fregistry.nlark.com%2Fmonaco-editor%2Fdownload%2Fmonaco-editor-0.24.0.tgz: ESOCKETTIMEDOUT".
#12 294.0 info If you think this is a bug, please open a bug report with the information provided in "/apisix-dashboard/web/yarn-error.log".
#12 294.0 info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
#12 294.0 make: *** [build] Error 1
------
Dockerfile.dashboard.rpm:24
--------------------
  23 |     # install APISIX dashboard
  24 | >>> RUN /install-common.sh install_dashboard \
  25 | >>>     # determine dist and write it into /tmp/dist file
  26 | >>>     && /determine-dist.sh
  27 |     
--------------------
error: failed to solve: process "/bin/sh -c /install-common.sh install_dashboard     && /determine-dist.sh" did not complete successfully: exit code: 2
make: *** [Makefile:124: build-dashboard-rpm] Error 1
Error: Process completed with exit code 2.
```

starsz commented on api/cmd/cache_verify.go:
```diff
@@ -0,0 +1,184 @@
+				return
+			}
+
+			url := fmt.Sprintf("http://%s:%d/apisix/admin/cache_verify", host, port)
```
Don't support https ?

RedemptionC replied:
@starsz 
do you mean that it's possible to use only https
and the http* url will become unavailable?
What do you think it's the right way here?

starsz replied:
Since the manager-api doesn't support only use https now.
So I think I will resolve this conversation.

starsz commented on api/internal/handler/cache_verify/cache_verify.go:
```diff
@@ -0,0 +1,220 @@
+	"context"
+	"encoding/json"
+	"fmt"
+	"github.com/apisix/manager-api/internal/conf"
```
We can move it later.

RedemptionC committed:
chore: runs goimport on this file
